ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/HD/DbChangelog.hs

```diff
@@ -0,0 +1,430 @@
+         (Ord k, Eq v)
+      => SeqDiffMK k v
+      -> (SeqDiffMK k v, SeqDiffMK k v)
+    splitSeqDiff (SeqDiffMK sq) =
+        bimap SeqDiffMK SeqDiffMK
+      $ splitAt (AS.length imm) sq
+
+    lr :: LedgerTables l (Product2 SeqDiffMK SeqDiffMK)
+    lr = mapLedgerTables (uncurry Pair2 . splitSeqDiff) changelogDiffs
+
+    l = mapLedgerTables (\(Pair2 x _) -> x) lr
+    r = mapLedgerTables (\(Pair2 _ y) -> y) lr
+
+    ldblog = DbChangelog {
+        changelogDiffAnchor
+      , changelogDiffs           = l
+      , changelogImmutableStates = imm
+      , changelogVolatileStates  = AS.Empty immTip
+      }
+
+    rdblog = DbChangelog {
+        changelogDiffAnchor      = getTipSlot (unDbChangelogState immTip)
+      , changelogDiffs           = r
+      , changelogImmutableStates = AS.Empty immTip
+      , changelogVolatileStates  = vol
+      }
+
+-- | Flush the immutable changes to the backing store
+--
+-- The 'Ouroboros.Consensus.Storage.ChainDB.Impl.LgrDB.LgrDb'
+-- 'Ouroboros.Consensus.Storage.ChainDB.Impl.LgrDB.flushLock' write lock must be
+-- held before calling this function.
+--
+-- PRECONDITION: @dblog@ should only contain the diffs for the immutable part
+-- of the changelog. If not, the @slot@ that we flush to the backing store will
+-- not match the actual tip of the diffs that we flush to the backing store.
+--
+-- == Note [Behaviour of flushIntoBackingStore]
+--
+-- Consider a simplified @DbChangelog@ that consists of a diff, an immutable
+-- part, and a volatile part. By /flush/ we mean @'flushIntoBackingStore'@.
+--
+-- >   changelog = diffs, imm, vol
+-- >   flush     = @'OnDisk.flush'@
+--
+-- Consider a changelog shaped as shown below, which spans blocks a, b, c, and d.
+--
+-- >   a     b     c     d
+-- >   |-----|     |-----|
+-- >     imm         vol
+-- >   |-----------------|
+-- >          diffs
+--
+-- Each block @x@ has an accompanying diff @diff(x)@. If we flush this
+-- changelog, the following will happen:
+-- * All diffs @diff(a) .. diff(d)@ will be flushed to the backing store.
+-- * The slot number of the tip of the immutable part, which is the slot
+--   number of block b, will be flushed to the backing store.
+--
+-- This leads to an inconsistency: We flushed all diffs up to and including
+-- the diff of block d, but we flushed the slot number of block b.
+--
+-- To counter this problem, we should either ...
+-- * Remove the diffs that are in the volatile part, or ...
+-- * Let the immutable part span blocks a..d, and leave the volatile part
+--   empty.
+--
+--           Solution 1                   Solution 2
+-- >    a     b     c     d           a     b     c     d
+-- >    |-----|     |-----|           |-----------------|  |---|
+-- >      imm         vol                     imm           vol
+-- >    |-----|                       |-----------------|
+-- >     diffs                               diffs
+--
+-- Which solution to use depends on the desired behaviour, which in turn
```

Comment:
It seems we need to update this comment. We already chose a solution, didn't we?